Design docs for the user API (#121)
CMakePP/CMaize: docs/src/developer/design/user_api.rst

      <name>
      SOURCE_DIR <where_the_source_files_are_located>
      INCLUDE_DIRS <directories_containing_header_files>
      DEPENDS <dependency0> <dependency1>
   )

Like the "Find Dependencies" step before it, the APIs for defining build targets
are designed primarily for collecting information pertaining to the build
target. Unlike the "Find Dependencies" step the backend of API calls for
defining build targets is CMake. The result of calling these methods are
properly configured CMake targets.

Test Project
============

After targets are built the next step is to test that they were built correctly.
Testing build targets with CMake often requires:

- finding dependencies of the testing framework,
- building testing targets which consume the project's build targets, and
- registering the tests with CTest.

It is also worth noting that tests are often built conditionally (e.g., a
:term:`build system` typically does not build the tests of dependencies built
during the "Find Dependencies" step). To that end CMake defines the
``BUILD_TESTING`` variable; when set to ``TRUE`` tests are built, otherwise they
are not. Proposed CMaize APIs for testing a project:

> .. code-block:: CMake

Reviewer: Are these mostly proposed interfaces that do not exist right now? I believe the only one we have is `cmaize_add_tests`, and CTest's `add_test` command is not necessary to register the test. That is done internally in `cmaize_add_tests`.

   cmaize_find_or_build_test_dependency(
       <name>
       URL <where_on_the_internet_to_download_from>
       VERSION <the_version_you_want>
       BUILD_TARGET <target_to_build>
       FIND_TARGET <target_representing_package>
       CMAKE_ARGS <configuration_options_to_set>
   )

>    cmaize_add_test_library(

Reviewer: What would a test library look like? Typically, we build an executable that is run to run the tests.

Reviewer: Or maybe the better question is, when would a test library be used?

Author: We get them when working with Python bindings. For example, to see if a Python object is properly converted to its C++ equivalent, you write a C++ library with a function that takes the C++ object, and checks its state. Then you call that function from your Python unit test. We also have them for some of the tutorials in PluginPlay as we need to create a dummy plugin to demonstrate how it works.

       <name>
       SOURCE_DIR <where_the_source_files_are_located>
       INCLUDE_DIRS <directories_containing_header_files>
       DEPENDS <dependency0> <dependency1> ...
   )

   cmaize_add_test_executable(
       <name>
       SOURCE_DIR <where_the_source_files_are_located>
       INCLUDE_DIRS <directories_containing_header_files>
       DEPENDS <dependency0> <dependency1> ...
   )

   # This is actually CTest's add_test command
   add_test(NAME <name> COMMAND)

   # This is a convenience function for the common scenario where the
   # add_test call simply calls the executable arising from the
   # cmaize_add_test_executable

>    cmaize_add_test(

Reviewer: Currently, `cmaize_add_test` is named `cmaize_add_tests` and is a wrapper around calling `cmaize_add_executable` and registering it to CTest with `add_test`.

Author: I'm advocating for renaming it to `cmaize_add_test` for consistency. (We can keep `cmaize_add_test` and forward it so people don't need to remember whether it's plural or not).

Reviewer: Sounds good, I am fine with either option.

       <name>
       SOURCE_DIR <where_the_source_files_are_located>
       INCLUDE_DIRS <directories_containing_header_files>
       DEPENDS <dependency0> <dependency1> ...
   )

All of the test functions are thin wrappers around the non-test functions of
the same name (e.g., ``cmaize_add_test_library`` wraps ``cmaize_add_library``),
which hide the logic for including the CTest CMake module, and checking that
``BUILD_TESTING`` is enabled (if it's not the functions are no-ops).

Install Project
===============

> If the tests are successful (or were skipped) it's on to :term:`package`

Reviewer: ```suggestion
If the tests are successful (or were skipped), the next step is :term:`package`
```

installation. Installation typically requires specifying which targets are

> part of the package, generating the packaging files, and then literally

Reviewer: ```suggestion
part of the package, generating the packaging files, and then
```

moving the targets and files to their final location. The main considerations
for installing are:

> - Collecting sufficient information to be able to install the package including:

Reviewer: Potentially use sub-bullets for the list in this item.

  where it goes, which pieces get installed, and what the runtime dependencies
  are.
- Installation should be done in a manner which considers the package manager.

The proposed installation :term:`API` is:

.. code-block:: CMake

>    cmaize_add_package(<name> NAMESPACE <namespace>)

Reviewer: Currently, `cmaize_add_package` can also accept a `VERSION` option and `TARGETS` list, at least for the CMakePackageManager version. `VERSION` specifies the version of the package while `TARGETS` specifies the list of targets to actually include in the package. Do you think either of these are something to include in the general API as well?

Author: `TARGETS` should eventually be exposed, but I don't really see a need to expose `VERSION`.

Each of the user API calls proceeding ``cmaize_add_package`` record the
information provided. In turn when it comes time to write the packaging files
and install the package, CMaize can do so in a largely automatic manner simply
by inspecting the information which was already provided. The only piece of
information which has not been been provided yet is the namespace to use in

> the package files (CMake recommends appending a prefix to installed targets to

Reviewer: Do you have a citation of CMake recommending using namespaces? It would be good to link to something for further reading into this and potentially as a gateway into other recommendations made by CMake.

After a quick search, I found some information [here](https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html) (search for "namespace" on the page), but I'm not sure if it is sufficient. Here is a [link](https://cmake.org/cmake/help/latest/policy/CMP0028.html) to the policy defining double colons (`::`) as a pattern used in namespacing `IMPORTED` and `ALIAS` targets.
